src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectReloadManager.cs

```diff
@@ -0,0 +1,508 @@
+        /// Removes the file change watch on the project file. 
+        /// </summary>
+        /// <returns></returns>
+        public async Task UnregisterProjectAsync(IReloadableProject project)
+        {
+            await _threadHandling.SwitchToUIThread();
+            
+            UnregisterProject(project);
+        }
+
+        private void UnregisterProject(IReloadableProject project)
+        {
+            uint filechangeCookie;
+            if (_registeredProjects.TryGetValue(project, out filechangeCookie))
+            {
+                // Remove watch
+                IVsFileChangeEx fileChangeService = _serviceProvider.GetService<IVsFileChangeEx, SVsFileChangeEx>();
+                if (fileChangeService != null)
+                {
+                    int hr = fileChangeService.UnadviseFileChange(filechangeCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr));
+                }
+
+                // Always remove the watcher from our list
+                _registeredProjects.Remove(project);
+            }
+        }
+
+        /// <summary>
+        /// Adds a file change watcher on the project file.
+        /// </summary>
+        private async Task ConnectToSolutionEvents()
+        {
+            await _threadHandling.SwitchToUIThread();
+
+            if (_solutionEventsCookie == VSConstants.VSCOOKIE_NIL)
+            {
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    int hr = solution.AdviseSolutionEvents(this, out _solutionEventsCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr) && _solutionEventsCookie != VSConstants.VSCOOKIE_NIL);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes the file change watch on the project file. 
+        /// </summary>
+        private async Task DisconnectFromSolutionEvents()
+        {
+            if(_solutionEventsCookie != VSConstants.VSCOOKIE_NIL)
+            {
+                await _threadHandling.SwitchToUIThread();
+
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    int hr = solution.UnadviseSolutionEvents(_solutionEventsCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr));
+                    _solutionEventsCookie = VSConstants.VSCOOKIE_NIL;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Handles one time initialization
+        /// </summary>
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            _reloadDelayScheduler = new TaskDelayScheduler(TimeSpan.FromMilliseconds(ReloadDelay), _threadHandling, CancellationToken.None);
```

Comment:
Is there any chance that this method could get called twice after the object initializes? We're doing a not null check below, should we be doing a null check here?